pkg/v1/layout/write.go

```diff
@@ -127,6 +130,102 @@ func (l Path) AppendDescriptor(desc v1.Descriptor) error {
 	return l.WriteFile("index.json", rawIndex, os.ModePerm)
 }
 
+// ReplaceImage writes a v1.Image to the Path and updates
+// the index.json to reference it, replacing any existing one that matches matcher, if found.
+func (l Path) ReplaceImage(img v1.Image, matcher match.Matcher, options ...Option) error {
+	if err := l.WriteImage(img); err != nil {
+		return err
+	}
+
+	desc, err := partial.Descriptor(img)
+	if err != nil {
+		return err
+	}
+
+	for _, opt := range options {
+		if err := opt(desc); err != nil {
+			return err
+		}
+	}
+
+	return l.ReplaceDescriptor(*desc, matcher)
+}
+
+// ReplaceIndex writes a v1.ImageIndex to the Path and updates
+// the index.json to reference it, replacing any existing one that matches matcher, if found.
+func (l Path) ReplaceIndex(ii v1.ImageIndex, matcher match.Matcher, options ...Option) error {
+	if err := l.WriteIndex(ii); err != nil {
+		return err
+	}
+
+	desc, err := partial.Descriptor(ii)
+	if err != nil {
+		return err
+	}
+
+	for _, opt := range options {
+		if err := opt(desc); err != nil {
+			return err
+		}
+	}
+
+	return l.ReplaceDescriptor(*desc, matcher)
+}
+
+// ReplaceDescriptor adds a descriptor to the index.json of the Path, replacing
+// any one matching matcher, if found.
+func (l Path) ReplaceDescriptor(desc v1.Descriptor, matcher match.Matcher) error {
```

Comment:
(Sorry I missed this before.)

Do you need this exposed? I don't think we need `mutate.AppendableDescriptor` unless this is exposed, since we can just pass through the index or image...

If you do need it, that's fine, I just don't want to expose anything that doesn't have an immediate need.

Reply:
You didn't miss it before. I had to add it. When you pass something to mutate.AppendManifests, it has to be a mutate.IndexAddendum, which has to have an Appendable, which _returns_ a Descriptor but not _actually_ a Descriptor. So I needed something to wrap a Descriptor if you actually have one. 

I did try just using IndexAddendum and passing the Descriptor with the Add blank. It gives a SIGSEGV. 

If you have a better approach, I'm all for it.

Reply:
I totally misunderstood your comment before, since I responded from my phone. Now that I read it, I understand what you are saying. I think you are right. I will try to rework it.

Reply:
So in retrospect, we _can_ remove `ReplaceDescriptor` and just have `ReplaceImage` and `ReplaceIndex`.

The only reason why I can think to keep it is because for append we have:

* `layout.AppendImage`
* `layout.AppendIndex`
* `layout.AppendDescriptor`

So it would make sense to keep it parallel with

* `layout.ReplaceImage`
* `layout.ReplaceIndex`
* `layout.ReplaceDescriptor`

Your call @jonjohnsonjr . If you want to keep it parallel, we can keep it as is. If you prefer not to have the `ReplaceDescriptor`, we can kill it.

My leaning is towards parallel, but I can go either way, so your call.

Then let's close this out.

Reply:
I am sympathetic towards the parallelism, but I'd lean towards killing it just so that we can get rid of `mutate.AppendableDescriptor`, and as you can probably tell, I bias heavily against exposing anything that we don't need to :P

Reply:
Done. Take a look.